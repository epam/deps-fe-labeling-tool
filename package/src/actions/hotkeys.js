import { createAction } from 'redux-actions'

const FEATURE_NAME = 'HOT_KEYS'

const registerHotKeyEvents = createAction(
  `${FEATURE_NAME}/ADD_HOT_KEY_EVENTS`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  registerHotKeyEvents,
  resetDefault
}
