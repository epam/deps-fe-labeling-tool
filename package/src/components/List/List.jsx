import React from 'react'
import { List as AntdList } from 'antd'
import 'antd/lib/list/style/index.less'
import { ListItem as StyledListItem, ListItemMeta as StyledListItemMeta } from './List.styles'

const List = (props) => (
  <AntdList {...props} />
)

const ListItem = (props) => (
  <StyledListItem {...props} />
)

const ListItemMeta = (props) => (
  <StyledListItemMeta {...props} />
)

export {
  List,
  ListItem,
  ListItemMeta
}
