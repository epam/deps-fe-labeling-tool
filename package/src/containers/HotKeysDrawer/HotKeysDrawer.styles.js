import styled from 'styled-components'
import { Column } from '@/components/Column'
import { Drawer } from '@/components/Drawer'
import { Row } from '@/components/Row'
import { Tabs } from '@/components/Tabs'
import { Tag } from '@/components/Tag'

const StyledTag = styled(Tag)`
  &.ant-tag {
    margin: 0.1rem 0.7rem;
  }
`
const StyledRow = styled(Row)`
  margin: 0.7rem;
`

const StyledCol = styled(Column)`
  margin: 0.3rem 0;
`

const StyledTab = styled(Tabs)`
  .ant-tabs-tab {
    cursor: default;
  }
`

const StyledDrawer = styled(Drawer)`
  .ant-drawer-header-title {
    flex: initial;
  }
`

export {
  StyledDrawer as Drawer,
  StyledTag as Tag,
  StyledRow as Row,
  StyledCol as Col,
  StyledTab as Tabs
}
