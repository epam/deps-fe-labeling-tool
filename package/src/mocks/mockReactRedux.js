const mockDispatch = jest.fn((action) => action)

const createMockMapStateToProps = (mapStateToProps) => {
  if (!mapStateToProps) {
    return mapStateToProps
  }

  return (state, ownProps) => ({
    props: mapStateToProps(state, ownProps)
  })
}

const getDispatchFunctionFormShorthandObject = (actionsObj) => (dispatch) => (
  Object.keys(actionsObj).reduce(
    (props, actionKey) => ({
      ...props,
      [actionKey]: jest.fn(
        (...args) => dispatch(actionsObj[actionKey].apply(undefined, args))
      )
    }),
    {}
  )
)

const createMockMapDispatchToProps = (mapDispatchToProps) => {
  if (!mapDispatchToProps) {
    return mapDispatchToProps
  }

  const mappingFunction = typeof mapDispatchToProps === 'function'
    ? mapDispatchToProps
    : getDispatchFunctionFormShorthandObject(mapDispatchToProps)

  return (dispatch, ownProps) => {
    dispatch = dispatch || mockDispatch

    return {
      props: mappingFunction(dispatch, ownProps),
      dispatch
    }
  }
}

const createMockMergeProps = (mergeProps) => {
  if (!mergeProps) {
    return mergeProps
  }

  return (stateProps, dispatchProps, ownProps) => {
    dispatchProps = dispatchProps || { dispatch: mockDispatch }

    return {
      props: mergeProps(stateProps, dispatchProps, ownProps),
      dispatch: dispatchProps.dispatch
    }
  }
}

const mockReactRedux = {
  connect: (mapStateToProps, mapDispatchToProps, mergeProps) => (WrappedComponent) => ({
    mapStateToProps: createMockMapStateToProps(mapStateToProps),
    mapDispatchToProps: createMockMapDispatchToProps(mapDispatchToProps),
    mergeProps: createMockMergeProps(mergeProps),
    WrappedComponent
  }),
  batch: (fn) => Promise.resolve(fn),
  useDispatch: jest.fn(() => mockDispatch),
  useSelector: jest.fn((selector) => selector())
}

export {
  mockReactRedux
}
