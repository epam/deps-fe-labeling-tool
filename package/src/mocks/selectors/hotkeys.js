
import { HotKeyEvent } from '@/constants/hotKeys'

const hotKeysSelector = jest.fn(() => [HotKeyEvent.COPY.name])

const mockHotKeysSelectors = {
  hotKeysSelector
}

export {
  mockHotKeysSelectors
}
