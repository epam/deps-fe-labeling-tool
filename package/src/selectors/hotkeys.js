const rootSelector = (state) => state.hotKeys

export {
  rootSelector as hotKeysSelector
}
