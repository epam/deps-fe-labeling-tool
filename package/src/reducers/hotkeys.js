import { handleActions } from 'redux-actions'
import { registerHotKeyEvents, resetDefault } from '@/actions/hotkeys'

const defaultState = []

const resetDefaultHandler = () => (defaultState)

const registerHotKeyEventsHandler = (state, action) => ([
  ...new Set(
    [
      ...state,
      ...action.payload
    ]
  )
])

const hotKeysReducer = handleActions(
  new Map([
    [
      registerHotKeyEvents,
      registerHotKeyEventsHandler
    ],
    [
      resetDefault,
      resetDefaultHandler
    ]
  ]),
  defaultState
)

export {
  hotKeysReducer
}
