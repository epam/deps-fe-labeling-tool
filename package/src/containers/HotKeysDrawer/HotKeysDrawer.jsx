import React, { useMemo } from 'react'
import { capitalize } from 'lodash'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { List, ListItem } from '@/components/List'
import { TabsSize } from '@/components/Tabs'
import { HotKeyEvent, HotKeyModifier } from '@/constants/hotKeys'
import { Placement } from '@/enums/Placement'
import { hotKeysSelector } from '@/selectors/hotkeys'
import {
  Tag,
  Row,
  Col,
  Tabs,
  Drawer
} from './HotKeysDrawer.styles'

const { TabPane } = Tabs

const DRAWER_HEIGHT = 350
const RESOURCE_HOTKEYS_INFO_TAB_KEY = 'info'
const RESOURCE_HOTKEYS_INFO_TAB_TITLE = 'Hot keys info'
const KEY_SEQUENCE_SEPARATOR = '+'
const WORDS_TO_DELETE_REGEXP = new RegExp('Key|Arrow', 'g')
const DRAWER_STYLE_PROPERTIES = {
  BODY: {
    paddingTop: 0
  },
  HEADER: {
    justifyContent: 'flex-end'
  }
}

const getUserFriendlyShortcuts = (shortcuts) => (
  shortcuts.map((shortcut) =>
    shortcut
      .replaceAll(WORDS_TO_DELETE_REGEXP, '')
      .split('+')
      .map(capitalize)
  )
)

const renderHotKeyElement = (key, index, array) => (
  <React.Fragment key={key}>
    <Tag>{key}</Tag>
    {index !== array.length - 1 && KEY_SEQUENCE_SEPARATOR}
  </React.Fragment>
)

const renderHotKey = (hotKey) => (
  <ListItem key={hotKey.description}>
    <Row>
      <Col span={8}>{hotKey.description}</Col>
      <Col span={16}>
        {
          getUserFriendlyShortcuts(hotKey.getShortcutsArray())
            .map((shortcut) => shortcut.map(renderHotKeyElement))
        }
        {hotKey.note && hotKey.note}
      </Col>
    </Row>
  </ListItem>
)

const HotKeysDrawer = ({ visible, toggleHotKeysInfo, hotKeyEvents }) => {
  const hotKeysInfo = useMemo(
    () => (
      hotKeyEvents
        .map((hotKeyEvent) => HotKeyEvent[hotKeyEvent] || HotKeyModifier[hotKeyEvent])
        .filter((hotKeyEvent) => !!hotKeyEvent.description)
    ),
    [hotKeyEvents]
  )
  return (
    <Drawer
      bodyStyle={DRAWER_STYLE_PROPERTIES.BODY}
      closable={false}
      headerStyle={DRAWER_STYLE_PROPERTIES.HEADER}
      height={DRAWER_HEIGHT}
      onClose={toggleHotKeysInfo}
      placement={Placement.BOTTOM}
      visible={visible}
    >
      <Tabs
        defaultActiveKey={RESOURCE_HOTKEYS_INFO_TAB_KEY}
        centered
        size={TabsSize.SMALL}
      >
        <TabPane
          tab={RESOURCE_HOTKEYS_INFO_TAB_TITLE}
          key={RESOURCE_HOTKEYS_INFO_TAB_KEY}
        >
          {
            hotKeysInfo.length && (
              <List
                grid={{ gutter: 18, column: 3 }}
                split={false}
                dataSource={hotKeysInfo}
                renderItem={renderHotKey}
              />
            )
          }
        </TabPane>
      </Tabs>
    </Drawer>
  )
}

HotKeysDrawer.propTypes = {
  visible: PropTypes.bool.isRequired,
  toggleHotKeysInfo: PropTypes.func.isRequired,
  hotKeyEvents: PropTypes.arrayOf(PropTypes.string).isRequired
}

const mapStateToProps = (state) => ({
  hotKeyEvents: hotKeysSelector(state)
})

const ConnectedComponent = connect(mapStateToProps)(HotKeysDrawer)

export { ConnectedComponent as HotKeysDrawer }
