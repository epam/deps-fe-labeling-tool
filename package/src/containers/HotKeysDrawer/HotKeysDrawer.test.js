import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockHotKeysSelectors } from '@/mocks/selectors/hotkeys'
import React from 'react'
import { shallow } from 'enzyme'
import { List } from '@/components/List'
import { Drawer } from './HotKeysDrawer.styles'
import { HotKeysDrawer } from '.'

jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/selectors/hotkeys', () => mockHotKeysSelectors)

const { WrappedComponent, mapStateToProps } = HotKeysDrawer

describe('Container: HotKeysDrawer', () => {
  describe('mapStateToProps', () => {
    it('should pass expected props to WrappedComponent', () => {
      const { props } = mapStateToProps()
      expect(props).toMatchSnapshot()
    })
  })

  let wrapper, defaultProps

  beforeEach(() => {
    defaultProps = {
      visible: true,
      toggleHotKeysInfo: jest.fn(),
      hotKeyEvents: mockHotKeysSelectors.hotKeysSelector()
    }

    wrapper = shallow(<WrappedComponent {...defaultProps} />)
  })

  it('should render correct layout', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot of renderItem with the selected tools', () => {
    const list = wrapper.find(List)
    const hotKeysInfo = list.props().dataSource
    const listItem = hotKeysInfo.map((info) => list.props().renderItem(info))
    const listWrapper = shallow(<div>{listItem}</div>)
    expect(listWrapper).toMatchSnapshot()
  })

  it('should call props toggleHotKeysInfo when calling to Drawer onClose', () => {
    wrapper.find(Drawer).props().onClose()
    expect(defaultProps.toggleHotKeysInfo).toHaveBeenCalled()
  })
})
