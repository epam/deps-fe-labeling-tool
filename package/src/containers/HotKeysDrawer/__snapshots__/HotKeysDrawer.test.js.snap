// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: HotKeysDrawer mapStateToProps should pass expected props to WrappedComponent 1`] = `
Object {
  "hotKeyEvents": Array [
    "COPY",
  ],
}
`;

exports[`Container: HotKeysDrawer should match snapshot of renderItem with the selected tools 1`] = `
<div>
  <ListItem
    key="Copy"
  >
    <HotKeysDrawerstyles__StyledRow>
      <HotKeysDrawerstyles__StyledCol
        span={8}
      >
        Copy
      </HotKeysDrawerstyles__StyledCol>
      <HotKeysDrawerstyles__StyledCol
        span={16}
      >
        <HotKeysDrawerstyles__StyledTag>
          Ctrl
        </HotKeysDrawerstyles__StyledTag>
        +
        <HotKeysDrawerstyles__StyledTag>
          C
        </HotKeysDrawerstyles__StyledTag>
      </HotKeysDrawerstyles__StyledCol>
    </HotKeysDrawerstyles__StyledRow>
  </ListItem>
</div>
`;

exports[`Container: HotKeysDrawer should render correct layout 1`] = `
<HotKeysDrawerstyles__StyledDrawer
  bodyStyle={
    Object {
      "paddingTop": 0,
    }
  }
  closable={false}
  headerStyle={
    Object {
      "justifyContent": "flex-end",
    }
  }
  height={350}
  onClose={[MockFunction]}
  placement="bottom"
  visible={true}
>
  <HotKeysDrawerstyles__StyledTab
    centered={true}
    defaultActiveKey="info"
    size="small"
  >
    <TabPane
      key="info"
      tab="Hot keys info"
    >
      <List
        dataSource={
          Array [
            HotKey {
              "description": "Copy",
              "name": "COPY",
              "note": undefined,
              "shortcut": Array [
                "ctrl+KeyC",
              ],
            },
          ]
        }
        grid={
          Object {
            "column": 3,
            "gutter": 18,
          }
        }
        renderItem={[Function]}
        split={false}
      />
    </TabPane>
  </HotKeysDrawerstyles__StyledTab>
</HotKeysDrawerstyles__StyledDrawer>
`;
